### ISSUE TITLE
Convolution reverb

### ISSUE THREAD

**Issue by blancoberg** [surge-synthesizer/surge#7817]
**Describe the solution you'd like:**
A convolution reverb in the effect lane. 
This has already been discussed in surge discord, but though I could post it here as well as a reminder. 

A convolution library that is potentially a candidate for this could be FFTConvolver:
https://github.com/HiFi-LoFi/FFTConvolver

**Comment by baconpaul** [comment]
The primary problem I’ve faced when thinking about this (because yes that fft convolver works well) is the workflow around saving loading managing and patch including the irs. (The second thing is ir sample rate mismatch but that’s fairly well understood). 

Writing a good version of that workflow is the primary thing that’s stopped me from doing a cr with this lib

**Comment by mkruselj** [comment]
A lot of that workflow can be factored with Shortcircuit's missing content handling I think.

My main takeaway here is that IRs should not be embedded into the patch file _by default_. They can be larger than wavetables. This can and will easily bloat the size of patch banks. And so there should be a factory IR path, a user IR path, and a way to _optionally_ embed IRs into the patch when saving (just like we do with tunings). Patches should also be able to reference the IRs through relative paths of course (especially in case of, say, no user IR path, relative path should then be taking the path of the patch as base).

But at that point, we might as well do the exact same thing with wavetables, as this would probably very much help in reducing the size of our factory bank (I bet a decent amount of patches share the same wavetables)...

**Comment by baconpaul** [comment]
I completely agree that if we had all of short-circuits missing resolution code available, finished, shared, and portable into surge, we could use that.

But that does make me wonder why you labeled this a Surge XT 1 milestone issue! :)

**Comment by mkruselj** [comment]
No specific XT1 release, easy to bump later as we see fit 😀

**Comment by blancoberg** [comment]
> No specific XT1 release, easy to bump later as we see fit 😀

Xt 1.4? 🥹

**Comment by mkruselj** [comment]
Are you volunteering? 😁

**Comment by mkruselj** [comment]
Here's something to look into!

https://github.com/HiFi-LoFi/KlangFalter

Could be nice to profile the performance of this one.

**Comment by blancoberg** [comment]
> Are you volunteering? 😁

If I could I would :) but I don't even know where to start tbh

**Comment by baconpaul** [comment]
https://github.com/Chowdhury-DSP/chowdsp_convolution

**Comment by Andreya-Autumn** [comment]
Oooooh. Nice!

**Comment by blancoberg** [comment]
> "Uniformly partitioned" doesn't sound like "zero latency" to me. In that regard,

Isn't that just referring to the supported block size? uniformly being 32, 64, 128 etc. So it shouldn't matter for surge if i understand it correctly? Although fftconvolver does seem to support both.

**Comment by jatinchowdhury18** [comment]
Essentially the way partitioning works is that you split up the IR into a bunch of "segments" (uniform just means all the segments are the same size). Then as you start processing input, you can build up a "history" of input segments to match the IR segments, and do FFT-based convolution on each segment. The simplest way to do this is to wait until all your input segments are full before computing the convolution, but that incurs 1 segment length of latency.

Alternatively, you can always convolve whatever signal is in the first input segment (which may not be full) with the first IR segment. This allows the processing to be zero-latency, but will use a bit more CPU, especially if your input is not always a power-of-2 number of samples.

 
At the moment the library supports both processing with zero-latency, or processing with 1 block of added latency as mentioned above.

**Comment by baconpaul** [comment]
Is block size 8 or 32 tractable or does it require larger blocks?

**Comment by baconpaul** [comment]
And thanks!

**Comment by mkruselj** [comment]
32 would be way better due to CPU usage.

**Comment by mx** [comment]
I didn't realize there was an open issue for this. I've been working on this effect, my goal is to have it out for the 1.4 release. I've implemented the saving/loading of IRs to patches, the browsing is gonna be pretty damn rough though.

**Comment by mkruselj** [comment]
I think, regarding browsing, indeed there should be something like what Wavetable and Window oscillators have, but since we don't have the waveform display area in the FX section (and I think it's not really a good time to add that to Surge right now), it should probably just look like a regular dropdown menu. No prev-next arrow buttons to leave more room for IR filename.

**Comment by Andreya-Autumn** [comment]
Alright. Starting the list of folks whom to credit for IRs. 

Jacob Gustavsson
John and Eleanor from Balance Mastering
